aioasuswrt/connection.py

```diff
@@ -13,59 +15,149 @@
 asyncssh.set_log_level("WARNING")
 
 
-class SshConnection:
+class _CommandException(Exception):
+    pass
+
+
+class _BaseConnection(abc.ABC):
+    def __init__(
+        self, host: str, port: int, username: Optional[str], password: Optional[str]
+    ):
+        self._host = host
+        self._port = port
+        self._username = username if username else None
+        self._password = password if password else None
+
+        self._io_lock = asyncio.Lock()
+
+    @property
+    def description(self) -> str:
+        """ Description of the connection."""
+        ret = f"{self._host}:{self._port}"
+        if self._username:
+            ret = f"{self._username}@{ret}"
+
+        return ret
+
+    async def async_run_command(self, command: str, retry=True) -> List[str]:
+        """ Call a command using the connection."""
+        async with self._io_lock:
+            if not self.is_connected:
+                await self.async_connect()
+
+            try:
+                return await self._async_call_command(command)
+            except _CommandException:
+                pass
+
+        # The command failed
+        if retry:
+            _LOGGER.debug(f"Retrying command: {command}")
+            return await self._async_call_command(command)
+        return []
+
+    async def async_connect(self):
+        if self.is_connected:
+            _LOGGER.debug(f"Connection already established to: {self.description}")
+            return
+
+        await self._async_connect()
+
+    async def async_disconnect(self):
```

Comment:
This PR contain a **breaking change**: `disconnect` method is replaced by `async_disconnect`.
I think it should be properly highlighted.

Reply:
Hmm yes might have been a bit too forward with deleting the previous disconnect method. Although I'm still of the opinion it should be removed, as the lock is definitely needed to avoid nasty errors.

Probably best to re-add the disconnect method including a deprecation warning, to at least give people some time for updating.

Reply:
Version upgrade using py is not mandatory, imho if breaking change is required must be implemented and released without deprecation or work-around, but is important to clarify to the users what is happening. Probably is better to insert in a mayor release as suggested by @kennedyshead  because breaking changes are expected.